src/Gui/PythonEditor.cpp

```diff
@@ -256,6 +265,69 @@ void PythonEditor::onUncomment()
     cursor.endEditBlock();
 }
 
+void PythonEditor::onExecuteInConsole()
+{
+    QTextCursor cursor = textCursor();
+    int selStart = cursor.selectionStart();
+    int selEnd = cursor.selectionEnd();
+    QTextBlock block;
+    QString selectedCode;
+
+    for (block = document()->begin(); block.isValid(); block = block.next()) {
+        int pos = block.position();
+        int off = block.length() - 1;
+        if (pos >= selStart || pos + off >= selStart) {
+            if (pos + 1 > selEnd)
+                break;
+
+            QString lineText = block.text();
+            selectedCode.append(lineText + QLatin1String("\n"));
+        }
+    }
+
+    if (!selectedCode.isEmpty()) {
+
+        /** Dedent the block of code so that the first selected
+         *  line has no indentation, but the remaining lines
+         *  keep their indentation relative to that first line.
+         */
+
+        // get the leading whitespace of the first line
+        QStringList lines = selectedCode.split(QLatin1Char('\n'));
+        QString firstLineIndent;
+        for (const QString& line : lines) {
+            if (!line.isEmpty()) {
+                int leadingWhitespace = line.indexOf(QRegularExpression(QLatin1String("\\S")));
+                if (leadingWhitespace > 0) {
+                    firstLineIndent = line.left(leadingWhitespace);
+                }
+                break;
+            }
+        }
+
+        // remove that first line whitespace from all the lines
+        for (QString& line : lines) {
+            if (!line.isEmpty() && line.startsWith(firstLineIndent)) {
+                line.remove(0, firstLineIndent.length());
+            }
+        }
+
+        // join the lines into a single QString so we can execute as a single block
+        QString dedentedCode = lines.join(QLatin1Char('\n'));

```

Comment:
This could be extracted into separate function also, then the code of this function would be simpler to follow as it would only contain logic for execution and not selecting the code and dedenting it.